markused regression: generic iterator fails with structs

### Describe the bug

I noticed that the test suite for my module failed with a C error without the `-no-skip-unused` option.

Notably, the test code works fine when run as a non-test, for example:

The failing test: https://github.com/gechandesu/ranges/blob/a05536043672c1529a7723c01a67f2c96b4314b7/ranges_test.v#L93-L121

Edited:
```v
// main.v
import ranges

struct Int {
        val int
}

fn (a Int) + (b Int) Int {
        return Int{a.val + b.val}
}

fn (a Int) - (b Int) Int {
        return Int{a.val - b.val}
}

fn (a Int) < (b Int) bool {
        return a.val < b.val
}

fn (a Int) == (b Int) bool {
        return a.val == b.val
}

fn main() {
        mut result := []Int{}
        start := Int{0}
        end := Int{5}
        step := Int{1}
        for i in ranges.range[Int](start, end, step) {
                result << i
        }
        assert result == [
                Int{0},
                Int{1},
                Int{2},
                Int{3},
                Int{4},
                Int{5},
        ]
        println('success')
}
```
```console
$ v run main.v
success
```

### Reproduction Steps

```
git clone https://github.com/gechandesu/ranges.git && cd ranges
git checkout a055360
v test .
```

### Expected Behavior

All tests passed.

### Current Behavior

```
---- Testing... ----------------------------------------------------------------------------------------------------------------------------
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/tsession_7f60ff978780_01KREGHWCQATS5YE037QVK1E6D/ranges_test.01KREGHXMQ8P3CN3SYSTTNG7HV.tmp.c:15233: warning: implicit declaration of function 'ranges__Range_T_math__big__Integer_next_T_math__big__Integer'
cc: /tmp/v_1000/tsession_7f60ff978780_01KREGHWCQATS5YE037QVK1E6D/ranges_test.01KREGHXMQ8P3CN3SYSTTNG7HV.tmp.c:15233: error: '{' expected (got ";")
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error:
==================
C error found while compiling generated C code.
This can be caused by invalid C interop code, C compiler flags, or a V compiler bug.
If your code is pure V and this still happens, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
FAIL     0.000 ms /home/ge/Code/v/ranges/ranges_test.v
>> compilation failed:
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/tsession_7f60ff978780_01KREGHWCQATS5YE037QVK1E6D/ranges_test.01KREGHXMQ8P3CN3SYSTTNG7HV.tmp.c:15233: warning: implicit declaration of function 'ranges__Range_T_math__big__Integer_next_T_math__big__Integer'
cc: /tmp/v_1000/tsession_7f60ff978780_01KREGHWCQATS5YE037QVK1E6D/ranges_test.01KREGHXMQ8P3CN3SYSTTNG7HV.tmp.c:15233: error: '{' expected (got ";")
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error:
==================
C error found while compiling generated C code.
This can be caused by invalid C interop code, C compiler flags, or a V compiler bug.
If your code is pure V and this still happens, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .

--------------------------------------------------------------------------------------------------------------------------------------------
To reproduce just failure 1 run:    '/home/ge/.vlang/v'  '/home/ge/Code/v/ranges/ranges_test.v'
Summary for all V _test.v files: 1 failed, 1 total. Elapsed time: 1862 ms, on 1 job. Comptime: 1806 ms. Runtime: 0 ms.
```

Failing tests: 
- https://github.com/gechandesu/ranges/blob/a05536043672c1529a7723c01a67f2c96b4314b7/ranges_test.v#L52-L68
- https://github.com/gechandesu/ranges/blob/a05536043672c1529a7723c01a67f2c96b4314b7/ranges_test.v#L93-L121

### Possible Solution

_No response_

### Additional Information/Context

`v -no-skip-unused test .` pass.

### V version

V 0.5.1 395b1ff

### Environment details (OS name and version, etc.)

|V full version      |V 0.5.1 2f3c0f1f397c08bfc5ef73e7ff62794471b2b811.395b1ff
|:-------------------|:-------------------
|OS                  |linux, Linux version 7.0.3-arch1-1 (linux@archlinux) (gcc (GCC) 15.2.1 20260209, GNU ld (GNU Binutils) 2.46) #1 SMP PREEMPT_DYNAMIC Thu, 30 Apr 2026 18:41:12 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |0.52GB/15.34GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2026-05-12 15:38:50
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/ge/Code/v/ranges
|                    |
|Git version         |git version 2.54.0
|V git status        |weekly.2026.03-1928-g395b1ff3
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20260209
|gcc version         |gcc (GCC) 15.2.1 20260209
|clang version       |clang version 22.1.3
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -802,9 +802,30 @@ pub fn (mut w Walker) stmt(node_ ast.Stmt) {
 					}
 				}
 				cond_type_sym := w.table.sym(resolved_cond_type)
-				if next_fn := cond_type_sym.find_method('next') {
+				mut next_method := cond_type_sym.find_method('next')
+				// Generic instantiated structs (e.g. `Range[big.Integer]`) have
+				// no methods of their own; the `next` method lives on the
+				// generic parent (`Range[T]`). The concrete types from the
+				// instantiation must be propagated to the walker, otherwise
+				// `remove_unused_fn_generic_types` may overwrite
+				// `fn_generic_types[next]` with the union derived from other
+				// callers (e.g. only `[int]` from a direct `iter.next()`),
+				// dropping the receiver type used here.
+				mut receiver_concrete_types := []ast.Type{}
+				if next_method == none && cond_type_sym.info is ast.Struct {
+					if cond_type_sym.info.concrete_types.len > 0
+						&& cond_type_sym.info.parent_type != 0 {
+						parent_sym := w.table.sym(cond_type_sym.info.parent_type)
+						if parent_next := parent_sym.find_method('next') {
+							next_method = parent_next
+							receiver_concrete_types = w.receiver_concrete_types(resolved_cond_type)
+						}
+					}
+				}
+				if next_fn := next_method {
 					unsafe {
-						w.fn_decl(mut &ast.FnDecl(next_fn.source_fn))
+						w.fn_decl_with_concrete_types(mut &ast.FnDecl(next_fn.source_fn),
+							receiver_concrete_types)
 					}
 				}
 			}

--- a/vlib/v/tests/skip_unused/generic_fn_instantiation_pruning_test.v
+++ b/vlib/v/tests/skip_unused/generic_fn_instantiation_pruning_test.v
@@ -175,3 +175,63 @@ fn test_skip_unused_marks_dependencies_inside_generic_anon_fns() {
 		panic(res.output)
 	}
 }
+
+fn test_skip_unused_keeps_generic_next_for_for_in_when_direct_calls_use_other_types() {
+	// Regression for https://github.com/vlang/v/issues/27147 .
+	// A `for in` over a generic iterator instantiation (e.g. `Range[big.Integer]`)
+	// must emit the matching `next` instantiation, even when other code calls
+	// methods on a different instantiation of the same generic struct directly
+	// (e.g. `iter.next()` on a `Range[int]`).
+	tmp_dir := os.join_path(os.vtmp_dir(), 'v_issue_27147')
+	os.mkdir_all(tmp_dir) or { panic(err) }
+	defer {
+		os.rmdir_all(tmp_dir) or {}
+	}
+	source_path := os.join_path(tmp_dir, 'issue_27147.v')
+	binary_path := os.join_path(tmp_dir, 'issue_27147')
+	source := [
+		'module main',
+		'',
+		'import math.big',
+		'',
+		'struct Range[T] {',
+		'\tstart T',
+		'\tend   T',
+		'\tstep  T',
+		'mut:',
+		'\tcur T',
+		'}',
+		'',
+		'pub fn (mut r Range[T]) next() ?T {',
+		'\tif r.cur > r.end {',
+		'\t\treturn none',
+		'\t}',
+		'\tdefer { r.cur += r.step }',
+		'\treturn r.cur',
+		'}',
+		'',
+		'pub fn (mut r Range[T]) reset() {',
+		'\tr.cur = r.start',
+		'}',
+		'',
+		'pub fn range[T](start T, end T, step T) Range[T] {',
+		'\treturn Range[T]{start: start, end: end, step: step, cur: start}',
+		'}',
+		'',
+		'fn main() {',
+		'\tmut iter := range(0, 5, 1)',
+		'\titer.reset()',
+		'\titer.next()',
+		'\tend := big.integer_from_int(3)',
+		'\tfor i in range[big.Integer](big.zero_int, end, big.one_int) {',
+		'\t\tprintln(i)',
+		'\t}',
+		'}',
+	].join('\n')
+	os.write_file(source_path, source) or { panic(err) }
+	res :=
+		os.execute('${os.quoted_path(vexe)} -o ${os.quoted_path(binary_path)} ${os.quoted_path(source_path)}')
+	if res.exit_code != 0 {
+		panic(res.output)
+	}
+}

